subsys/usb/device_next/class/usbd_cdc_acm.c

```diff
@@ -549,6 +549,12 @@ static void cdc_acm_tx_fifo_handler(struct k_work *work)
 	len = ring_buf_get(data->tx_fifo.rb, buf->data, buf->size);
 	net_buf_add(buf, len);
 
+	if (IS_ENABLED(CONFIG_USBD_CDC_ACM_ENSURE_SHORT_PKT_FOR_LAST_TX_PKT)) {
+		if ((len != 0) && ((len % cdc_acm_get_bulk_mps(c_data)) == 0)) {
+			udc_ep_buf_set_zlp(buf);
```

Comment:
This will have a performance impact, because this forces ZLP even if new data is written to the ring buffer during the IN transfer. What should be done, in my opinion is enqueueing:
  * all the data if len < mps, OR
  * ROUND_DOWN(len, mps) if len >= mps

Then after the request is finished, if the ring buffer is empty and last transmitted buffer was mps multiple (which can be stored in some boolean flag), deliberately send ZLP.

Reply:
This is actually the idea of the initial code when this PR was raised. Which was kind of abandoned after the discussion that took place, to better utilize the existing ZLP flag for endpoints in UDC. 

Now I am wondering which path we should go here, as we are having different opinions on this: 
1. favor performance like what you have suggested
2. Simply go for the simpler handling that matches the existing handling. 
I personally favor performance in this regard.

Reply:
The ZLP can be added together with improved performance (even when the ZLP option is not selected). The existing handling has various other issues (e.g. #79104) and here it seems that the solution would be simplest if based on top of #79104.

Reply:
@tmon-nordic, It seems to be ideal for me that : 
1. #79104 use `ring_buffer_is_empty()` instead of `tx_fifo.altered` flags to check whether there are more transaction needed. Then, 
2. For this PR, we add one `ZLP needed` flag to indicate the current transmitting packet might need ZLP, and send ZLP out if tx_fifo is still empty by the time the current IN transaction completed. 

Shall we wait for @jfischer-no  input on these 2 items? 
@jfischer-no are you okay with the handling above for these 2 PR? If so I will make the changes accordingly, and we will be able to move forward. 

Thank you.

Reply:
I cannot really think of why this plan wouldn't work. I think you can just proceed to implementation without waiting for anyone's permission.

Reply:
Let me look for time to do the change and verification before pushing the change up.

Reply:
Hi @tmon-nordic, 
Changes discussed above made and tested working at my side. Please review again and see whether its all good. 
Thank you.

Reply:
This PR will inevitably conflict with #79104.

Reply:
Yeah @tmon-nordic , this will inevitably conflict with #79104. So I am planning to rebase either one of the PR once the other one make it into main branch. Is that the correct flow to go in this kind of scenario?

Reply:
Hi @tmon-nordic, Since #78160 (that replaces #79104) is merged in, I have resolved the merge conflict and pushed the change. Please have a look again and see whether its alright with you. 😉

Reply:
At High-Speed this already gives the best throughput because `cdc_acm_ep_pool` returns 512 bytes large buffers. At Full-Speed the `len = ring_buf_get(data->tx_fifo.rb, buf->data, buf->size);` would need to be changed according to https://github.com/zephyrproject-rtos/zephyr/pull/79193#discussion_r1787233847 to get the best throughput.

We might eventually get to reduce the `cdc_acm_ep_pool` buffer for Full-Speed only builds, so this can probably stay as is for now.

Reply:
IMO, from both of the options above, I think making cdc_acm_ep_pool size adjustable is the better option as it helps 
- Reduce memory footprint in the case where the buffer needed is only 64 bytes for Full-Speed only compilation. 
- If current shim/hal/controller for a chip could support autonomous transfer of multiple packets, it might give better performance as it could do the multiple packet transfer with lesser overhead. 

I suggest we keep it as is for now 😄